pinot-controller/src/main/java/org/apache/pinot/controller/LeadControllerManager.java

```diff
@@ -0,0 +1,163 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.controller;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.helix.model.ExternalView;
+import org.apache.pinot.common.config.TableNameBuilder;
+import org.apache.pinot.common.utils.helix.LeadControllerUtils;
+import org.apache.pinot.controller.helix.core.PinotHelixResourceManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.pinot.common.utils.CommonConstants.Helix.LEAD_CONTROLLER_RESOURCE_NAME;
+import static org.apache.pinot.common.utils.CommonConstants.Helix.NUMBER_OF_PARTITIONS_IN_LEAD_CONTROLLER_RESOURCE;
+
+
+public class LeadControllerManager {
+  private static final Logger LOGGER = LoggerFactory.getLogger(LeadControllerManager.class);
+
+  private Map<Integer, Integer> _partitionIndexCache;
+  private PinotHelixResourceManager _pinotHelixResourceManager;
```

Comment:
We seem to have a mutual dependency between PinotHelxResourceManager and this class. Can we break that?

Reply:
The purpose of bringing `PinotHelxResourceManager` here is to fetch the EV of leadControllerResource, where all the requirements like `KeyBuilder` and `DataAccessor` live.
And `PinotHelxResourceManager` needs `LeadControllerManager` to register the cache for MasterSlaveStateModelFactory.
So it's hard to break that. That's why I added a method `registerResourceManager()` in `LeadControllerManager`. One thing we can do is add an `unregisterResourceManager()` in stop method.

Reply:
Refactored the code to pass `HelixManager` to the constructor of `LeadControllerManager`. No need to have methods to register/unregister `PinotHelxResourceManager`.